DivideByZeroException with concurrent initialisation of TermInfoDriver

## Steps to Reproduce

1. Create an empty .net console app targeting net461
2. Attempt to access any of the lockable properties from TermInfoDriver on multiple threads (example property I'm using is Background/ForegroundColor: https://github.com/mono/mono/blob/cf24b9f9d4b80512122f998a47cae850df1d4cb2/mcs/class/corlib/System/TermInfoDriver.cs#L403-L435)

#### Sample code to show exception
```
class Program
    {
		public static void Main(string[] args)
		{
			Parallel.Invoke(
				() => { var x = Console.ForegroundColor; },
				() => { Console.BackgroundColor = ConsoleColor.Black; }
			);
		}
    }
```

3. Build and run the application

### Current Behavior

Occasionally, the application will run without an issue.  However, most of the time an exception is thrown due to the behaviour I've described in this other thread [here](https://github.com/akkadotnet/akka.net/issues/2432#issuecomment-568130295).

### Expected Behavior

No exception should be thrown.

## On which platforms did you notice this

[x] macOS
[x] Linux
[x] Windows

**Version Used**:

6.4.0

## Stacktrace

```
Unhandled Exception:
System.AggregateException: One or more errors occurred. ---> System.DivideByZeroException: Attempted to divide by zero.
  at System.TermInfoDriver.ChangeColor (System.String format, System.ConsoleColor color) [0x00016] in <filename unknown>:0 
  at System.TermInfoDriver.set_BackgroundColor (System.ConsoleColor value) [0x00011] in <filename unknown>:0 
  at System.ConsoleDriver.set_BackgroundColor (System.ConsoleColor value) [0x0001f] in <filename unknown>:0 
  at System.Console.set_BackgroundColor (System.ConsoleColor value) [0x00000] in <filename unknown>:0 
  at akka_test.Program+<>c.<Main>b__0_1 () [0x00001] in <filename unknown>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x00012] in <filename unknown>:0 
  at System.Threading.Tasks.Task.Execute () [0x00016] in <filename unknown>:0 
[MVID] 2320d006b1a64388aa1812d52f51bde3 4
[MVID] 7747cf446af449e194e4b4e70d85e773 0,1,2,3,5,6
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.FastWaitAll (System.Threading.Tasks.Task[] tasks) [0x0006e] in <filename unknown>:0 
  at System.Threading.Tasks.Parallel.Invoke (System.Threading.Tasks.ParallelOptions parallelOptions, System.Action[] actions) [0x0032e] in <filename unknown>:0 
[MVID] 7747cf446af449e194e4b4e70d85e773 0,1
---> (Inner Exception #0) System.DivideByZeroException: Attempted to divide by zero.
  at System.TermInfoDriver.ChangeColor (System.String format, System.ConsoleColor color) [0x00016] in <filename unknown>:0 
  at System.TermInfoDriver.set_BackgroundColor (System.ConsoleColor value) [0x00011] in <filename unknown>:0 
  at System.ConsoleDriver.set_BackgroundColor (System.ConsoleColor value) [0x0001f] in <filename unknown>:0 
  at System.Console.set_BackgroundColor (System.ConsoleColor value) [0x00000] in <filename unknown>:0 
  at akka_test.Program+<>c.<Main>b__0_1 () [0x00001] in <filename unknown>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x00012] in <filename unknown>:0 
  at System.Threading.Tasks.Task.Execute () [0x00016] in <filename unknown>:0 
[MVID] 2320d006b1a64388aa1812d52f51bde3 4
[MVID] 7747cf446af449e194e4b4e70d85e773 0,1,2,3,5,6<---

[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> System.DivideByZeroException: Attempted to divide by zero.
  at System.TermInfoDriver.ChangeColor (System.String format, System.ConsoleColor color) [0x00016] in <filename unknown>:0 
  at System.TermInfoDriver.set_BackgroundColor (System.ConsoleColor value) [0x00011] in <filename unknown>:0 
  at System.ConsoleDriver.set_BackgroundColor (System.ConsoleColor value) [0x0001f] in <filename unknown>:0 
  at System.Console.set_BackgroundColor (System.ConsoleColor value) [0x00000] in <filename unknown>:0 
  at akka_test.Program+<>c.<Main>b__0_1 () [0x00001] in <filename unknown>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x00012] in <filename unknown>:0 
  at System.Threading.Tasks.Task.Execute () [0x00016] in <filename unknown>:0 
[MVID] 2320d006b1a64388aa1812d52f51bde3 4
[MVID] 7747cf446af449e194e4b4e70d85e773 0,1,2,3,5,6
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.FastWaitAll (System.Threading.Tasks.Task[] tasks) [0x0006e] in <filename unknown>:0 
  at System.Threading.Tasks.Parallel.Invoke (System.Threading.Tasks.ParallelOptions parallelOptions, System.Action[] actions) [0x0032e] in <filename unknown>:0 
[MVID] 7747cf446af449e194e4b4e70d85e773 0,1
---> (Inner Exception #0) System.DivideByZeroException: Attempted to divide by zero.
  at System.TermInfoDriver.ChangeColor (System.String format, System.ConsoleColor color) [0x00016] in <filename unknown>:0 
  at System.TermInfoDriver.set_BackgroundColor (System.ConsoleColor value) [0x00011] in <filename unknown>:0 
  at System.ConsoleDriver.set_BackgroundColor (System.ConsoleColor value) [0x0001f] in <filename unknown>:0 
  at System.Console.set_BackgroundColor (System.ConsoleColor value) [0x00000] in <filename unknown>:0 
  at akka_test.Program+<>c.<Main>b__0_1 () [0x00001] in <filename unknown>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x00012] in <filename unknown>:0 
  at System.Threading.Tasks.Task.Execute () [0x00016] in <filename unknown>:0 
[MVID] 2320d006b1a64388aa1812d52f51bde3 4
[MVID] 7747cf446af449e194e4b4e70d85e773 0,1,2,3,5,6<---


Process finished with exit code 1.

```

## Notes

The core problem is that the lock on an uninitialised TermInfoDriver happens immediately upon accessing a property (get or set) and is released immediately before the initialisation process is complete.  So, subsequent calls which happen concurrently will complete before any expected fields are present.

I'm not sure about the best way to fix this but it would be good to have this lock released *after* the initialisation process happens, but we would need to potentially try/catch the whole init block to account for exceptions.  

<!--
You can join us at https://gitter.im/mono/mono to discuss your reported issue
-->

--- a/mcs/class/corlib/System/TermInfoDriver.cs
+++ b/mcs/class/corlib/System/TermInfoDriver.cs
@@ -196,76 +196,80 @@ namespace System {
 				if (inited)
 					return;
 				
-				/* This should not happen any more, since it is checked for in Console */
-				if (!ConsoleDriver.IsConsole)
-					throw new IOException ("Not a tty.");
-				
-				ConsoleDriver.SetEcho (false);
-				
-				string endString = null;
-				keypadXmit = reader.Get (TermInfoStrings.KeypadXmit);
-				keypadLocal = reader.Get (TermInfoStrings.KeypadLocal);
-				if (keypadXmit != null) {
-					WriteConsole (keypadXmit); // Needed to get the arrows working
-					if (keypadLocal != null)
-						endString += keypadLocal;
-				}
-				
-				origPair = reader.Get (TermInfoStrings.OrigPair);
-				origColors = reader.Get (TermInfoStrings.OrigColors);
-				setfgcolor = reader.Get (TermInfoStrings.SetAForeground);
-				setbgcolor = reader.Get (TermInfoStrings.SetABackground);
-				maxColors = reader.Get (TermInfoNumbers.MaxColors);
-				maxColors = Math.Max (Math.Min (maxColors, 16), 1);
-				
-				string resetColors = (origColors == null) ? origPair : origColors;
-				if (resetColors != null)
-					endString += resetColors;
-				
-				unsafe {
-					if (!ConsoleDriver.TtySetup (keypadXmit, endString, out control_characters, out native_terminal_size)){
-						control_characters = new byte [17];
-						native_terminal_size = null;
-						//throw new IOException ("Error initializing terminal.");
+				try {
+					/* This should not happen any more, since it is checked for in Console */
+					if (!ConsoleDriver.IsConsole)
+						throw new IOException ("Not a tty.");
+					
+					ConsoleDriver.SetEcho (false);
+					
+					string endString = null;
+					keypadXmit = reader.Get (TermInfoStrings.KeypadXmit);
+					keypadLocal = reader.Get (TermInfoStrings.KeypadLocal);
+					if (keypadXmit != null) {
+						WriteConsole (keypadXmit); // Needed to get the arrows working
+						if (keypadLocal != null)
+							endString += keypadLocal;
 					}
-				}
-				
-				stdin = new StreamReader (Console.OpenStandardInput (0), Console.InputEncoding);
-				clear = reader.Get (TermInfoStrings.ClearScreen);
-				bell = reader.Get (TermInfoStrings.Bell);
-				if (clear == null) {
-					clear = reader.Get (TermInfoStrings.CursorHome);
-					clear += reader.Get (TermInfoStrings.ClrEos);
-				}
-				
-				csrVisible = reader.Get (TermInfoStrings.CursorNormal);
-				if (csrVisible == null)
-					csrVisible = reader.Get (TermInfoStrings.CursorVisible);
-				
-				csrInvisible = reader.Get (TermInfoStrings.CursorInvisible);
-				if (term == "cygwin" || term == "linux" || (term != null && term.StartsWith ("xterm")) ||
-				    term == "rxvt" || term == "dtterm") {
-					titleFormat = "\x1b]0;{0}\x7"; // icon + window title
-				} else if (term == "iris-ansi") {
-					titleFormat = "\x1bP1.y{0}\x1b\\"; // not tested
-				} else if (term == "sun-cmd") {
-					titleFormat = "\x1b]l{0}\x1b\\"; // not tested
-				}
-				
-				cursorAddress = reader.Get (TermInfoStrings.CursorAddress);
-				
-				GetCursorPosition ();
-#if DEBUG
-				logger.WriteLine ("noGetPosition: {0} left: {1} top: {2}", noGetPosition, cursorLeft, cursorTop);
-				logger.Flush ();
-#endif
-				if (noGetPosition) {
-					WriteConsole (clear);
-					cursorLeft = 0;
-					cursorTop = 0;
+					
+					origPair = reader.Get (TermInfoStrings.OrigPair);
+					origColors = reader.Get (TermInfoStrings.OrigColors);
+					setfgcolor = reader.Get (TermInfoStrings.SetAForeground);
+					setbgcolor = reader.Get (TermInfoStrings.SetABackground);
+					maxColors = reader.Get (TermInfoNumbers.MaxColors);
+					maxColors = Math.Max (Math.Min (maxColors, 16), 1);
+					
+					string resetColors = (origColors == null) ? origPair : origColors;
+					if (resetColors != null)
+						endString += resetColors;
+					
+					unsafe {
+						if (!ConsoleDriver.TtySetup (keypadXmit, endString, out control_characters, out native_terminal_size)){
+							control_characters = new byte [17];
+							native_terminal_size = null;
+							//throw new IOException ("Error initializing terminal.");
+						}
+					}
+					
+					stdin = new StreamReader (Console.OpenStandardInput (0), Console.InputEncoding);
+					clear = reader.Get (TermInfoStrings.ClearScreen);
+					bell = reader.Get (TermInfoStrings.Bell);
+					if (clear == null) {
+						clear = reader.Get (TermInfoStrings.CursorHome);
+						clear += reader.Get (TermInfoStrings.ClrEos);
+					}
+					
+					csrVisible = reader.Get (TermInfoStrings.CursorNormal);
+					if (csrVisible == null)
+						csrVisible = reader.Get (TermInfoStrings.CursorVisible);
+					
+					csrInvisible = reader.Get (TermInfoStrings.CursorInvisible);
+					if (term == "cygwin" || term == "linux" || (term != null && term.StartsWith ("xterm")) ||
+						term == "rxvt" || term == "dtterm") {
+						titleFormat = "\x1b]0;{0}\x7"; // icon + window title
+					} else if (term == "iris-ansi") {
+						titleFormat = "\x1bP1.y{0}\x1b\\"; // not tested
+					} else if (term == "sun-cmd") {
+						titleFormat = "\x1b]l{0}\x1b\\"; // not tested
+					}
+					
+					cursorAddress = reader.Get (TermInfoStrings.CursorAddress);
+					
+					GetCursorPosition ();
+	#if DEBUG
+					logger.WriteLine ("noGetPosition: {0} left: {1} top: {2}", noGetPosition, cursorLeft, cursorTop);
+					logger.Flush ();
+	#endif
+					if (noGetPosition) {
+						WriteConsole (clear);
+						cursorLeft = 0;
+						cursorTop = 0;
+					}
+
+				} finally {
+					inited = true;
 				}
 
-				inited = true;
 			}
 		}
 

--- a/mcs/class/corlib/System/TermInfoDriver.cs
+++ b/mcs/class/corlib/System/TermInfoDriver.cs
@@ -195,7 +195,6 @@ namespace System {
 			lock (initLock){
 				if (inited)
 					return;
-				inited = true;
 				
 				/* This should not happen any more, since it is checked for in Console */
 				if (!ConsoleDriver.IsConsole)
@@ -265,6 +264,8 @@ namespace System {
 					cursorLeft = 0;
 					cursorTop = 0;
 				}
+
+				inited = true;
 			}
 		}
 

